endpoints/data_objects/src/main.cpp

```diff
@@ -1790,17 +1805,84 @@ namespace
 						}
 						else if (iter->second == "update_existing") {
 							opts = fs::copy_options::update_existing;
+							addKeyVal(&input.destDataObjInp.condInput, FORCE_FLAG_KW, "");
 						}
 						else if (iter->second != "none") {
+							log::error("{}: Invalid value for [option] parameter.");
 							return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
 						}
 					}
 
+					const fs::path from = src_lpath_iter->second;
+					const fs::path to = dst_lpath_iter->second;
+
+					fs::throw_if_path_length_exceeds_limit(from);
+					fs::throw_if_path_length_exceeds_limit(to);
+
 					auto conn = irods::get_connection(client_info.username);
-					const auto copied =
-						fs::client::copy_data_object(conn, src_lpath_iter->second, dst_lpath_iter->second, opts);
 
-					res.body() = json{{"irods_response", {{"status_code", 0}}}, {"copied", copied}}.dump();
+					if (!fs::client::is_data_object(conn, from)) {
+						throw fs::filesystem_error{
+							"path does not point to a data object",
+							from,
+							irods::experimental::make_error_code(INVALID_OBJECT_TYPE)};
+					}
+
+					bool copied = false;
+					bool execute_operation = true;
+
+					if (const auto s = fs::client::status(conn, to); fs::client::exists(s)) {
+						if (fs::client::equivalent(conn, from, to)) {
+							throw fs::filesystem_error{
+								"paths cannot point to the same object",
+								from,
+								to,
+								irods::experimental::make_error_code(SAME_SRC_DEST_PATHS_ERR)};
+						}
+
+						if (!fs::client::is_data_object(s)) {
+							throw fs::filesystem_error{
+								"path does not point to a data object",
+								to,
+								irods::experimental::make_error_code(INVALID_OBJECT_TYPE)};
+						}
+
+						if (fs::copy_options::skip_existing == opts) {
+							execute_operation = false;
+						}
+						else if (fs::copy_options::overwrite_existing == opts) {
+							addKeyVal(&input.destDataObjInp.condInput, FORCE_FLAG_KW, "");
+						}
+						else if (fs::copy_options::update_existing == opts) {
+							if (fs::client::last_write_time(conn, from) <= fs::client::last_write_time(conn, to)) {
+								execute_operation = false;
+							}
```

Comment:
Good question. `last_write_time` is supposed to return the mtime of the latest good replica. Because it is a precondition, I don't know that it matters that the `resource` targeted by the user matches the replica selected by `last_write_time`.

Reply:
Yes. That's what I assumed one would use `update_existing` for... *Please* tell me if I completely missed the point here. :p

I guess I'll add a note and say for the above hypothetical that we can assume the user has permission to overwrite the object, included the force flag, etc.

Reply:
Oh, I see. I never included *what* the copy action was. Yes, the client's intent is to copy object0 over object1. Sorry.

Reply:
What you describe is correct.

`update_existing` results in a call to the copy API if and only if the src is newer than the **existing** data object.

Reply:
Now if `last_write_time` does not have a good replica, it may throw an exception. I think that's a bad idea and so I think the HTTP API needs to simply expose the copy API in its rawest form.

The HTTP API must model the iRODS API, not the filesystem library. That means we need to be careful of how we use the filesystem lib.

Thoughts?

Reply:
All of the filesystem logic for this op has been scrapped for the C API.

This makes the impl much simpler and aligns with the goal of the project better.